File: v4/rust/iprange-livedb/src/recovery/external_sort.rs
//! Bounded two-file merge sort for recovery-readable direct ranges.

use std::fs::File;
use std::mem::size_of;

use crate::cancellation::CancellationToken;
use crate::contract::MetaV4;
use crate::error::{Error, Result};
use crate::key::IpKey;
use crate::range_tree::Record;
use crate::validation::ValidationReason;

use super::direct_output::DirectKey;
use super::page_set::PageSet;
use super::range_scan::{self, RangeEvents};
use super::scratch::{residue_error, Scratch, ScratchCleanup, ScratchSlot};
use super::RecoveryBudget;

#[path = "external_sort/streams.rs"]
mod streams;
use streams::{merge_runs, read_run, record_order, write_run, RunReader};

pub(crate) struct ExternalSortFailure {
    pub(crate) cause: Error,
    pub(crate) cleanup: Option<ScratchCleanup>,
}

#[derive(Clone, Copy)]
pub(crate) struct SortRequest<'a> {
    pub(crate) meta: MetaV4,
    pub(crate) budget: &'a RecoveryBudget,
    pub(crate) retained_heap_bytes: u64,
    pub(crate) readable_records: u64,
    pub(crate) cancellation: &'a CancellationToken,
}

// Exact scratch ownership must remain in the error without a failure-path box.
#[allow(clippy::result_large_err)]
pub(crate) fn sort_and_emit<K: DirectKey>(
    file: &File,
    request: SortRequest<'_>,
    mut pages: PageSet,
    mut emit: impl FnMut(Record<K>) -> Result<()>,
) -> std::result::Result<ScratchCleanup, ExternalSortFailure> {
    let mut records =
        match prepare_records::<K>(request.budget, request.retained_heap_bytes, &pages) {
            Ok(records) => records,
            Err(cause) => return Err(page_failure(pages, cause)),
        };
    if let Err(cause) = pages.reset() {
        return Err(page_failure(pages, cause));
    }
    let mut scratch = match pages.take_scratch() {
        Some(scratch) => scratch,
        None => match start_scratch(request.meta, request.budget) {
            Ok(scratch) => scratch,
            Err(cause) => return Err(page_failure(pages, cause)),
        },
    };
    if let Err(cause) = scratch.require_external_sort() {
        let _ = pages.release(&mut scratch);
        return finish(scratch, Err(cause));
    }
    let first = match scratch.create() {
        Ok(first) => first,
        Err(cause) => {
            let _ = pages.release(&mut scratch);
            return finish(scratch, Err(cause));
        }
    };
    let runs = scan_runs(
        file,
        request.meta,
        &mut pages,
        &mut records,
        first,
        request.readable_records,
        request.cancellation,
        &mut scratch,
    );
    drop(records);
    let reusable = pages.release(&mut scratch);
    let result = runs.and_then(|runs| {
        sort_runs_and_emit(
            &mut scratch,
            runs,
            reusable,
            request.readable_records,
            request.cancellation,
            &mut emit,
        )
    });
    finish(scratch, result)
}

fn prepare_records<K: IpKey>(
    budget: &RecoveryBudget,
    retained_heap_bytes: u64,
    pages: &PageSet,
) -> Result<Vec<Record<K>>> {
    let record_size = size_of::<Record<K>>() as u64;
    let available = budget
        .max_heap_bytes
        .checked_sub(retained_heap_bytes)
        .and_then(|value| value.checked_sub(pages.retained_bytes()))
        .and_then(|value| value.checked_sub(record_size))
        .ok_or(Error::BudgetExceeded("recovery unordered range buffer"))?;
    let capacity = usize::try_from((available + record_size) / record_size)
        .unwrap_or(usize::MAX)
        .max(1);
    let mut records = Vec::new();
    records
        .try_reserve_exact(capacity)
        .map_err(|_| Error::BudgetExceeded("recovery unordered range buffer"))?;
    let retained = (records.capacity() as u64)
        .checked_mul(record_size)
        .ok_or(Error::ArithmeticOverflow("recovery unordered range buffer"))?;
    if retained > available + record_size {
        return Err(Error::BudgetExceeded("recovery unordered range buffer"));
    }
    Ok(records)
}

#[allow(clippy::too_many_arguments)]
fn scan_runs<K: DirectKey>(
    file: &File,
    meta: MetaV4,
    pages: &mut PageSet,
    records: &mut Vec<Record<K>>,
    first: ScratchSlot,
    readable_records: u64,
    cancellation: &CancellationToken,
    scratch: &mut Scratch,
) -> Result<Runs> {
    let mut runs = Runs::new(first);
    let capacity = records.capacity();
    {
        let mut events = ScanEvents {
            records,
            capacity,
            scratch,
            runs: &mut runs,
            cancellation,
            seen: 0,
        };
        range_scan::scan(file, meta, pages, cancellation, &mut events)?;
        events.flush()?;
        if events.seen != readable_records {
            return Err(Error::RecoveryCandidateChanged);
        }
    }
    Ok(runs)
}

fn sort_runs_and_emit<K: DirectKey>(
    scratch: &mut Scratch,
    runs: Runs,
    reusable: Option<ScratchSlot>,
    readable_records: u64,
    cancellation: &CancellationToken,
    emit: &mut impl FnMut(Record<K>) -> Result<()>,
) -> Result<()> {
    if readable_records == 0 {
        return Ok(());
    }
    let sorted = merge_all::<K>(scratch, runs, reusable, cancellation)?;
    emit_sorted(scratch, sorted, readable_records, cancellation, emit)
}

#[derive(Clone, Copy)]
struct Runs {
    slot: ScratchSlot,
    end: u64,
    count: u64,
}

impl Runs {
    fn new(slot: ScratchSlot) -> Self {
        Self {
            slot,
            end: 128,
            count: 0,
        }
    }

    fn append<K: DirectKey>(
        &mut self,
        scratch: &mut Scratch,
        records: &mut [Record<K>],
    ) -> Result<()> {
        if records.is_empty() {
            return Ok(());
        }
        records.sort_unstable_by(record_order);
        self.end = write_run(scratch, self.slot, self.end, records)?;
        self.count = self
            .count
            .checked_add(1)
            .ok_or(Error::ArithmeticOverflow("recovery scratch runs"))?;
        Ok(())
    }
}

struct ScanEvents<'a, K> {
    records: &'a mut Vec<Record<K>>,
    capacity: usize,
    scratch: &'a mut Scratch,
    runs: &'a mut Runs,
    cancellation: &'a CancellationToken,
    seen: u64,
}

impl<K: DirectKey> ScanEvents<'_, K> {
    fn flush(&mut self) -> Result<()> {
        self.runs.append(self.scratch, self.records)?;
        self.records.clear();
        Ok(())
    }
}

impl<K: DirectKey> RangeEvents<K> for ScanEvents<'_, K> {
    fn page_accepted(&mut self) -> Result<()> {
        Ok(())
    }

    fn page_rejected(&mut self, _io_unreadable: bool) -> Result<()> {
        Ok(())
    }

    fn unknown(
        &mut self,
        _reason: ValidationReason,
        _page: Option<u32>,
        _unbounded: bool,
    ) -> Result<()> {
        Ok(())
    }

    fn range(&mut self, _page: u32, record: Option<Record<K>>) -> Result<()> {
        let Some(record) = record else {
            return Ok(());
        };
        self.cancellation.check()?;
        self.records.push(record);
        self.seen = self
            .seen
            .checked_add(1)
            .ok_or(Error::ArithmeticOverflow("recovery readable ranges"))?;
        if self.records.len() == self.capacity {
            self.flush()?;
        }
        Ok(())
    }
}

fn merge_all<K: DirectKey>(
    scratch: &mut Scratch,
    mut input: Runs,
    reusable: Option<ScratchSlot>,
    cancellation: &CancellationToken,
) -> Result<ScratchSlot> {
    if input.count <= 1 {
        return Ok(input.slot);
    }
    let mut output = match reusable {
        Some(slot) => {
            scratch.reset(slot)?;
            slot
        }
        None => scratch.create()?,
    };
    while input.count > 1 {
        scratch.reset(output)?;
        let merged = merge_pass::<K>(scratch, input, output, cancellation)?;
        output = input.slot;
        input = merged;
    }
    Ok(input.slot)
}

fn merge_pass<K: DirectKey>(
    scratch: &mut Scratch,
    input: Runs,
    output: ScratchSlot,
    cancellation: &CancellationToken,
) -> Result<Runs> {
    let mut source_at = 128;
    let mut destination_at = 128;
    let mut remaining_runs = input.count;
    let mut output_runs = 0;
    while remaining_runs != 0 {
        cancellation.check()?;
        let left = read_run::<K>(scratch, input.slot, source_at)?;
        source_at = left.end;
        let right = if remaining_runs > 1 {
            let run = read_run::<K>(scratch, input.slot, source_at)?;
            source_at = run.end;
            Some(run)
        } else {
            None
        };
        destination_at = merge_runs::<K>(
            scratch,
            input.slot,
            output,
            destination_at,
            left,
            right,
            cancellation,
        )?;
        remaining_runs -= if right.is_some() { 2 } else { 1 };
        output_runs += 1;
    }
    if source_at != scratch.length(input.slot) {
        return Err(Error::Corrupt("scratch run framing has trailing bytes"));
    }
    Ok(Runs {
        slot: output,
        end: destination_at,
        count: output_runs,
    })
}

fn emit_sorted<K: DirectKey>(
    scratch: &Scratch,
    slot: ScratchSlot,
    expected: u64,
    cancellation: &CancellationToken,
    emit: &mut impl FnMut(Record<K>) -> Result<()>,
) -> Result<()> {
    let run = read_run::<K>(scratch, slot, 128)?;
    if run.count != expected || run.end != scratch.length(slot) {
        return Err(Error::Corrupt("final recovery scratch run is incomplete"));
    }
    let mut reader = RunReader::<K>::new(slot, run);
    while let Some(record) = reader.next(scratch)? {
        cancellation.check()?;
        emit(record)?;
    }
    Ok(())
}

#[allow(clippy::result_large_err)]
fn finish(
    scratch: Scratch,
    result: Result<()>,
) -> std::result::Result<ScratchCleanup, ExternalSortFailure> {
    let cleanup = scratch.cleanup();
    match (result, cleanup.clean()) {
        (Ok(()), true) => Ok(cleanup),
        (Err(cause), true) => Err(ExternalSortFailure {
            cause,
            cleanup: Some(cleanup),
        }),
        (result, false) => {
            let cleanup_error = residue_error(&cleanup);
            let cause = Error::CleanupIncomplete {
                cause: Box::new(
                    result.unwrap_err_or(Error::Corrupt("recovery scratch cleanup is incomplete")),
                ),
                cleanup: Box::new(cleanup_error),
            };
            Err(ExternalSortFailure {
                cause,
                cleanup: Some(cleanup),
            })
        }
    }
}

fn start_scratch(meta: MetaV4, budget: &RecoveryBudget) -> Result<Scratch> {
    let directory = budget
        .scratch_directory
        .as_deref()
        .ok_or(Error::BudgetExceeded("recovery unordered ranges"))?;
    Scratch::start(
        directory,
        meta,
        budget.max_scratch_bytes,
        budget.max_scratch_files,
        budget.max_open_files,
    )
}

#[allow(clippy::result_large_err)]
fn page_failure(pages: PageSet, cause: Error) -> ExternalSortFailure {
    match pages.finish(Err(cause)) {
        Err(failure) => ExternalSortFailure {
            cause: failure.cause,
            cleanup: failure.cleanup,
        },
        Ok(_) => unreachable!("failed page scan cannot finish successfully"),
    }
}

trait ResultExt {
    fn unwrap_err_or(self, fallback: Error) -> Error;
}

impl ResultExt for Result<()> {
    fn unwrap_err_or(self, fallback: Error) -> Error {
        match self {
            Ok(()) => fallback,
            Err(cause) => cause,
        }
    }
}

File: v4/go/internal/recovery/external_sort.go
package recovery

// Bounded two-file merge sort for recovery-readable direct ranges
// (Rust recovery/external_sort.rs): the readable records stream into
// sorted runs inside the heap record buffer, runs live in the mapped
// scratch files, and fixed passes merge down to one run which streams
// into the output. Every budget refusal and corruption class is the
// Rust class; the scratch cleanup folds through the same terminal.

import (
	"slices"

	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/live"
	"github.com/firehol/iprange/v4/go/internal/mapping"
	"github.com/firehol/iprange/v4/go/internal/validation"
)

// externalSortFailure is the failing terminal of one external sort
// (Rust ExternalSortFailure: the cause and the attached scratch
// cleanup).
type externalSortFailure struct {
	cause   error
	cleanup *scratchCleanup
}

// sortRequest is the bounded external-sort request (Rust SortRequest).
type sortRequest struct {
	meta              format.Meta
	budget            *RecoveryBudget
	retainedHeapBytes uint64
	readableRecords   uint64
	check             func() error
	initialArea       *sortArea
}

// sortArea is one scratch region usable as a sort area (Rust
// SortArea).
type sortArea struct {
	slot scratchSlot
	base uint64
}

// newSortArea builds one sort area at the payload base (Rust
// SortArea::new).
func newSortArea(slot scratchSlot, base uint64) sortArea {
	return sortArea{slot: slot, base: base}
}

// sortAndEmit runs the full external sort and streams the sorted
// records into the emit function (Rust sort_and_emit): the bounded
// record buffer, the reset page set, the scratch attempt (reused from
// the page set or started fresh), the run scan, and the merge pass
// chain fold through the scratch cleanup terminal.
func sortAndEmit(codec rangeCodec, m *mapping.Mapping, request sortRequest, pages *pageSet, emit func(rangeRecord) error) (*scratchCleanup, *externalSortFailure) {
	records, err := prepareRecords(codec, request.budget, request.retainedHeapBytes, pages)
	if err != nil {
		return nil, pageFailure(pages, err)
	}
	if err := pages.reset(); err != nil {
		return nil, pageFailure(pages, err)
	}
	scratch := pages.takeScratch()
	if scratch == nil {
		scratch, err = startSortScratch(request.meta, request.budget)
		if err != nil {
			return nil, pageFailure(pages, err)
		}
	}
	if err := scratch.requireExternalSort(); err != nil {
		pages.release(scratch)
		return finishSort(scratch, err)
	}
	first, err := firstSortArea(scratch, request.initialArea)
	if err != nil {
		pages.release(scratch)
		return finishSort(scratch, err)
	}
	workspace := newSortWorkspace()
	var runs runs
	scanErr := scanSortRuns(workspace, codec, m, request.meta, pages, &records, first, request.readableRecords, request.check, scratch, &runs)
	records = nil
	reusable := pages.release(scratch)
	var result error
	if scanErr == nil {
		result = sortRunsAndEmit(workspace, codec, scratch, runs, reusable, request.readableRecords, request.check, emit)
	} else {
		result = scanErr
	}
	return finishSort(scratch, result)
}

// firstSortArea selects the initial sort area (Rust first_area: the
// reuse area from the tables storage or a fresh scratch file, sized
// to the payload base).
func firstSortArea(scratch *scratch, initial *sortArea) (sortArea, error) {
	area := sortArea{}
	if initial != nil {
		area = *initial
	} else {
		slot, err := scratch.create()
		if err != nil {
			return sortArea{}, err
		}
		area = sortArea{slot: slot, base: scratchHeaderSize}
	}
	if err := scratch.resize(area.slot, area.base); err != nil {
		return sortArea{}, err
	}
	return area, nil
}

// prepareRecords allocates the exact bounded record buffer (Rust
// prepare_records: the heap minus the retained facts, at least one
// record, and the retained capacity proof).
func prepareRecords(codec rangeCodec, budget *RecoveryBudget, retainedHeapBytes uint64, pages *pageSet) ([]rangeRecord, error) {
	recordSize := uint64(codec.recordSize())
	available, ok := checkedSub(budget.MaxHeapBytes, retainedHeapBytes)
	if !ok {
		return nil, budgetError("recovery unordered range buffer")
	}
	available, ok = checkedSub(available, pages.retainedBytes())
	if !ok {
		return nil, budgetError("recovery unordered range buffer")
	}
	available, ok = checkedSub(available, recordSize)
	if !ok {
		return nil, budgetError("recovery unordered range buffer")
	}
	capacity := available / recordSize
	capacity++
	if capacity > uint64(maxInt) {
		capacity = uint64(maxInt)
	}
	records := make([]rangeRecord, 0, int(capacity))
	retained := uint64(cap(records)) * recordSize
	if retained > available+recordSize {
		return nil, budgetError("recovery unordered range buffer")
	}
	return records, nil
}

// scanSortRuns scans the mapped source into sorted runs (Rust
// scan_runs): every readable record joins the record buffer, the
// buffer flushes as a sorted run at capacity, and the final flush
// plus the readable-record proof close the scan.
func scanSortRuns(workspace *sortWorkspace, codec rangeCodec, m *mapping.Mapping, meta format.Meta, pages *pageSet, records *[]rangeRecord, first sortArea, readableRecords uint64, check func() error, scratch *scratch, out *runs) error {
	*out = runs{area: first, end: first.base}
	capacity := cap(*records)
	events := &scanSortEvents{
		codec:     codec,
		workspace: workspace,
		records:   records,
		capacity:  capacity,
		scratch:   scratch,
		runs:      out,
		check:     check,
	}
	if err := scanRanges(codec, m, meta, pages, check, events); err != nil {
		return err
	}
	if err := events.flush(); err != nil {
		return err
	}
	if events.seen != readableRecords {
		return candidateChangedError()
	}
	return nil
}

// sortRunsAndEmit merges the scanned runs and streams the final run
// (Rust sort_runs_and_emit: a zero-record source emits nothing).
func sortRunsAndEmit(workspace *sortWorkspace, codec rangeCodec, scratch *scratch, input runs, reusable *scratchSlot, readableRecords uint64, check func() error, emit func(rangeRecord) error) error {
	if readableRecords == 0 {
		return nil
	}
	sorted, err := mergeAllSortRuns(workspace, codec, scratch, input, reusable, check)
	if err != nil {
		return err
	}
	return emitSorted(workspace, codec, scratch, sorted, readableRecords, check, emit)
}

// runs is the run-set state of one merge pass (Rust Runs).
type runs struct {
	area  sortArea
	end   uint64
	count uint64
}

// appendRun sorts the buffered records and frames them as one run
// (Rust Runs::append).
func appendRun(r *runs, workspace *sortWorkspace, codec rangeCodec, scratch *scratch, records []rangeRecord) error {
	if len(records) == 0 {
		return nil
	}
	slices.SortFunc(records, func(left, right rangeRecord) int {
		switch {
		case lessRecord(codec, left, right):
			return -1
		case lessRecord(codec, right, left):
			return 1
		default:
			return 0
		}
	})
	end, err := writeRun(workspace, scratch, codec, r.area.slot, r.end, records)
	if err != nil {
		return err
	}
	r.end = end
	next := r.count + 1
	if next == 0 {
		return &format.Error{Code: format.CodeArithmeticOverflow, Detail: "recovery scratch runs"}
	}
	r.count = next
	return nil
}

// scanSortEvents consumes the range scan of one external sort (Rust
// ScanEvents).
type scanSortEvents struct {
	codec     rangeCodec
	workspace *sortWorkspace
	records   *[]rangeRecord
	capacity  int
	scratch   *scratch
	runs      *runs
	check     func() error
	seen      uint64
}

// flush sorts and frames the buffer as one run (Rust ScanEvents::flush).
func (e *scanSortEvents) flush() error {
	if err := appendRun(e.runs, e.workspace, e.codec, e.scratch, *e.records); err != nil {
		return err
	}
	*e.records = (*e.records)[:0]
	return nil
}

func (e *scanSortEvents) pageAccepted() error { return nil }
func (e *scanSortEvents) pageRejected(bool) error {
	return nil
}
func (e *scanSortEvents) unknown(reason validation.ValidationReason, page *uint32, unbounded bool) error {
	return nil
}
func (e *scanSortEvents) rangeEvent(page uint32, record rangeRecordOption) error {
	if !record.ok {
		return nil
	}
	if err := live.Checkpoint(e.check); err != nil {
		return err
	}
	*e.records = append(*e.records, record.value)
	next := e.seen + 1
	if next == 0 {
		return &format.Error{Code: format.CodeArithmeticOverflow, Detail: "recovery readable ranges"}
	}
	e.seen = next
	if len(*e.records) == e.capacity {
		return e.flush()
	}
	return nil
}

// mergeAllSortRuns reduces the run set to one run (Rust merge_all):
// the reusable file slot or a fresh scratch file becomes the output
// area, and every pass swaps the areas until one run remains.
func mergeAllSortRuns(workspace *sortWorkspace, codec rangeCodec, scratch *scratch, input runs, reusable *scratchSlot, check func() error) (sortArea, error) {
	if input.count <= 1 {
		return input.area, nil
	}
	output := sortArea{base: scratchHeaderSize}
	if reusable != nil {
		output.slot = *reusable
	} else {
		slot, err := scratch.create()
		if err != nil {
			return sortArea{}, err
		}
		output.slot = slot
	}
	for input.count > 1 {
		if err := scratch.resize(output.slot, output.base); err != nil {
			return sortArea{}, err
		}
		merged, err := mergeSortPass(workspace, codec, scratch, input, output, check)
		if err != nil {
			return sortArea{}, err
		}
		output = input.area
		input = merged
	}
	return input.area, nil
}

// mergeSortPass merges run pairs of the input area into the output
// area (Rust merge_pass): each merge consumes one or two runs, and
// the pass proves the input was consumed exactly.
func mergeSortPass(workspace *sortWorkspace, codec rangeCodec, scratch *scratch, input runs, output sortArea, check func() error) (runs, error) {
	sourceAt := input.area.base
	destinationAt := output.base
	remainingRuns := input.count
	outputRuns := uint64(0)
	for remainingRuns != 0 {
		if err := live.Checkpoint(check); err != nil {
			return runs{}, err
		}
		left, err := readRun(scratch, codec, input.area.slot, sourceAt)
		if err != nil {
			return runs{}, err
		}
		sourceAt = left.end
		var right *run
		if remainingRuns > 1 {
			nextRun, err := readRun(scratch, codec, input.area.slot, sourceAt)
			if err != nil {
				return runs{}, err
			}
			sourceAt = nextRun.end
			right = &nextRun
		}
		destinationAt, err = mergeRuns(workspace, scratch, codec, input.area.slot, output.slot, destinationAt, left, right, check)
		if err != nil {
			return runs{}, err
		}
		if right != nil {
			remainingRuns -= 2
		} else {
			remainingRuns--
		}
		outputRuns++
	}
	if sourceAt != scratch.length(input.area.slot) {
		return runs{}, &format.Error{Code: format.CodeFormatInvalid, Detail: "scratch run framing has trailing bytes"}
	}
	return runs{area: output, end: destinationAt, count: outputRuns}, nil
}

// emitSorted streams the final sorted run into the output (Rust
// emit_sorted): the run must carry exactly the readable records and
// end at the file length.
func emitSorted(workspace *sortWorkspace, codec rangeCodec, scratch *scratch, area sortArea, expected uint64, check func() error, emit func(rangeRecord) error) error {
	run, err := readRun(scratch, codec, area.slot, area.base)
	if err != nil {
		return err
	}
	if run.count != expected || run.end != scratch.length(area.slot) {
		return &format.Error{Code: format.CodeFormatInvalid, Detail: "final recovery scratch run is incomplete"}
	}
	reader := newRunReader(codec, area.slot, run, workspace.left[:])
	for {
		record, err := reader.next(scratch)
		if err != nil {
			return err
		}
		if !record.ok {
			return nil
		}
		if err := live.Checkpoint(check); err != nil {
			return err
		}
		if err := emit(record.value); err != nil {
			return err
		}
	}
}

// finishSort folds the sort result through the scratch cleanup
// terminal (Rust finish: a clean cleanup with a successful result
// returns the cleanup, any failure attaches it, and an unclean
// cleanup becomes the CleanupIncomplete class).
func finishSort(scratch *scratch, result error) (*scratchCleanup, *externalSortFailure) {
	cleanup := scratch.cleanup()
	clean := cleanup.clean()
	switch {
	case result == nil && clean:
		return cleanup, nil
	case !clean:
		cause := cleanupIncompleteError(result, cleanup)
		return nil, &externalSortFailure{cause: cause, cleanup: cleanup}
	default:
		return nil, &externalSortFailure{cause: result, cleanup: cleanup}
	}
}

// startSortScratch starts the scratch attempt of the external sort
// (Rust start_scratch).
func startSortScratch(meta format.Meta, budget *RecoveryBudget) (*scratch, error) {
	if budget.ScratchDirectory == "" {
		return nil, budgetError("recovery unordered ranges")
	}
	return scratchStart(budget.ScratchDirectory, meta, budget.MaxScratchBytes, budget.MaxScratchFiles, budget.MaxOpenFiles)
}

// pageFailure folds one pre-sort failure through the page-set
// terminal (Rust page_failure).
func pageFailure(pages *pageSet, cause error) *externalSortFailure {
	failure := pages.finish(cause)
	return &externalSortFailure{cause: failure.cause, cleanup: failure.cleanup}
}
